include/pybind11/detail/function_view.h

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2021 The Pybind Development Team.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "common.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename Signature>
+class function_view;
+
+template <typename R, typename... Arg>
+class function_view<R(Arg...)> {
+public:
+    function_view() = default;
+
+    /// Creates a view that references an existing function.
+    ///
+    /// The constructed view is only valid for the lifetime of `f`.
+    ///
+    /// Requires that `F` is a function object type that is callable with
+    /// arguments `(Args...)` and has a return type convertible to `R`.
+    template <typename F,
+              typename = enable_if_t<
+                  // Prevent construction from the same `function_view` type, to avoid inefficient
+                  // double wrapping that also commonly leads to dangling references.
+                  (!std::is_same<remove_cv_t<remove_reference_t<F>>, function_view>::value
```

Comment:
Any reason not to use std::decay here?

Reply:
That is fine, it just does extra conversions that aren't applicable, like array decaying and function reference decaying.  I think remove_cvref more precisely reflects what is needed here.